plugins/inputs/directory_monitor/directory_monitor.go

```diff
@@ -0,0 +1,468 @@
+## Warning: setting this number too high can lead to a high number of dropped metrics.
+# max_buffered_metrics = 1000
+#
+## The maximum amount of files to process at once. A very high number can lead to bigger memory use and potential file system errors.
+# max_concurrent_files = 3000
+#
+## The dataformat to be read from the files.
+## Each data format has its own unique set of configuration options, read
+## more about them here:
+## https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_INPUT.md
+## NOTE: We do not currently support JSON file streaming and thus JSON files will be fully loaded into memory when they are processed.
+data_format = "influx"
+`
+
+var (
+	defaultFilesToMonitor             = []string{}
+	defaultFilesToIgnore              = []string{}
+	defaultMaxBufferedMetrics         = 1000
+	defaultMaxConcurrentFiles         = 3000
+	defaultMonitorInterval            = internal.Duration{Duration: 50 * time.Millisecond}
+	defaultDirectoryDurationThreshold = internal.Duration{Duration: 50 * time.Millisecond}
+)
+
+type empty struct{}
+type semaphore chan empty
+
+type DirectoryMonitor struct {
+	Directory         string `toml:"directory"`
+	FinishedDirectory string `toml:"finished_directory"`
+	UseErrorDirectory bool   `toml:"use_error_directory"`
+	ErrorDirectory    string `toml:"error_directory"`
+
+	CharacterEncoding          string            `toml:"character_encoding"`
+	FilesToMonitor             []string          `toml:"files_to_monitor"`
+	FilesToIgnore              []string          `toml:"files_to_ignore"`
+	MaxBufferedMetrics         int               `toml:"max_buffered_metrics"`
+	MonitorInterval            internal.Duration `toml:"monitor_interval"`
+	DirectoryDurationThreshold internal.Duration `toml:"directory_duration_threshold"`
+	MaxConcurrentFiles         int               `toml:"max_concurrent_files"`
+
+	filesInUse          cmap.ConcurrentMap
+	Log                 telegraf.Logger
+	parserFunc          parsers.ParserFunc
+	decoder             *encoding.Decoder
+	filesProcessed      selfstat.Stat
+	filesDropped        selfstat.Stat
+	waitGroup           *sync.WaitGroup
+	acc                 telegraf.TrackingAccumulator
+	sem                 semaphore
+	quit                chan bool
+	fileRegexesToMatch  []*regexp.Regexp
+	fileRegexesToIgnore []*regexp.Regexp
+}
+
+func (monitor *DirectoryMonitor) SampleConfig() string {
+	return sampleConfig
+}
+
+func (monitor *DirectoryMonitor) Description() string {
+	return "Ingests files in a directory and then moves them to a target directory."
+}
+
+func (monitor *DirectoryMonitor) Start(acc telegraf.Accumulator) error {
+	// Use tracking to determine when more metrics can be added without overflowing the outputs.
+	monitor.acc = acc.WithTracking(monitor.MaxBufferedMetrics)
+	go func() {
+		for {
+			select {
+			case <-monitor.acc.Delivered():
+				<-monitor.sem
+			}
+		}
```

Comment:
range works really well with channels and closes cleanly
```suggestion
		for range monitor.acc.Delivered() {
				<-monitor.sem
		}
```